diff -uNr qemu-6.1.0/slirp/src/slirp.c qemu-6.1.0.mod/slirp/src/slirp.c
--- qemu-6.1.0/slirp/src/slirp.c	2021-08-25 22:41:18.897068615 +0300
+++ qemu-6.1.0.mod/slirp/src/slirp.c	2021-08-26 00:20:55.545515218 +0300
@@ -254,65 +254,49 @@
                                     socklen_t addrlen, uint32_t *scope_id,
                                     unsigned *cached_time)
 {
-    char buff[512];
-    char buff2[257];
-    FILE *f;
-    int found = 0;
-    union {
-        struct in_addr dns_addr;
-        struct in6_addr dns6_addr;
-    } tmp_addr;
+    void *tmp_addr = alloca(addrlen);
+    char *env_dns_addr = getenv("CONFIG_QEMU_DNS");
+    int found = 1;
     unsigned if_index;
 
-    assert(sizeof(tmp_addr) >= addrlen);
-    char *prefixed_resolvconf_path = android_dynamic_path_prefix(RESOLV_CONF_PATH);
-    f = fopen(prefixed_resolvconf_path, "r");
-    if (!f) {
-        free(prefixed_resolvconf_path);
-        return -1;
+    if (!env_dns_addr) {
+        env_dns_addr = "1.1.1.1";
     }
-    free(prefixed_resolvconf_path);
-
-    DEBUG_MISC("IP address of your DNS(s):");
-    while (fgets(buff, 512, f) != NULL) {
-        if (sscanf(buff, "nameserver%*[ \t]%256s", buff2) == 1) {
-            char *c = strchr(buff2, '%');
-            if (c) {
-                if_index = if_nametoindex(c + 1);
-                *c = '\0';
-            } else {
-                if_index = 0;
-            }
 
-            if (!inet_pton(af, buff2, &tmp_addr)) {
-                continue;
-            }
-            /* If it's the first one, set it to dns_addr */
-            if (!found) {
-                memcpy(pdns_addr, &tmp_addr, addrlen);
-                memcpy(cached_addr, &tmp_addr, addrlen);
-                if (scope_id) {
-                    *scope_id = if_index;
-                }
-                *cached_time = curtime;
+    if (!inet_pton(af, env_dns_addr, tmp_addr)) {
+        if (af == AF_INET) {
+            if (!inet_pton(AF_INET, "1.1.1.1", tmp_addr)) {
+                found = 0;
             }
-
-            if (++found > 3) {
-                DEBUG_MISC("  (more)");
-                break;
-            } else if (slirp_debug & DBG_MISC) {
-                char s[INET6_ADDRSTRLEN];
-                const char *res = inet_ntop(af, &tmp_addr, s, sizeof(s));
-                if (!res) {
-                    res = "  (string conversion error)";
-                }
-                DEBUG_MISC("  %s", res);
+        } else {
+            if (!inet_pton(AF_INET6, "2606:4700:4700::1111", tmp_addr)) {
+                found = 0;
             }
         }
     }
-    fclose(f);
-    if (!found)
+
+    char *c = strchr(env_dns_addr, '%');
+
+    if (c) {
+        if_index = if_nametoindex(c + 1);
+        *c = '\0';
+    } else {
+        if_index = 0;
+    }
+
+    memcpy(pdns_addr, tmp_addr, addrlen);
+    memcpy(cached_addr, tmp_addr, addrlen);
+
+    if (scope_id) {
+        *scope_id = if_index;
+    }
+
+    *cached_time = curtime;
+
+    if (!found) {
         return -1;
+    }
+
     return 0;
 }
 
